### DISCUSSION TITLE
twi.h

### DISCUSSION THREAD

**Discussion by GPappy1959** [collin80/SavvyCAN#596] [Q&A]
I have been unsuccessfully trying to get my new EVTV ESP32 CANDUe running for a few days.  I followed the instructions that I found in the readme.md file in GVRET and found a post on the Savvycan.com website that got me most of the way.  Unfortunately I keep getting an error when I try to compile the file.  

fatal error: include/twi.h: No such file or directory

I searched for information on this error and did find some discussion but was wondering where I went wrong in the setup that this file is not being loaded properly.

Thanks in advance for any thoughts.

**Comment by collin80** [comment]
Usually when people get an error like that it means they aren't compiling for the proper board. One thing in particular that stands out is that you mentioned GVRET but also that you have an ESP32 CANDue. They aren't compatible. This should probably have been made more clear. You need ESP32RET for that board. It's found here: https://github.com/collin80/ESP32RET and has its own set of required libraries and files in order to compile it. The required parts are all listed in its readme.

Alternatively, basically all of the boards have prebuilt binaries if all you want to do is use the board.
For an ESP32 board the prebuilt file is this one: https://savvycan.com/ESP32RET_Updater.zip

Just extract those files, plug the board in, and run the relevant batch file (.bat for windows, .sh for linux, .command for MacOS) and it should place the proper binary onto the board for you. This will be a lot easier than trying to compile it yourself. But, if you had hoped to write your own software and/or make changes to the ESP32RET program then you'd need to follow the readme and get all the files set up.

**Reply by GPappy1959** [reply]
Thanks Collin for responding.  I literally just figured out what was going on.  I had brought in both the SdFat and SdFat-beta libraries but only renamed the SdFat library.  I was rereading your notes and saw that it said that you were now using the beta version so I removed the Master from it and added master back onto the Sd-fat libraries and it loaded.  

Now I get to go back and do it again with the correct files.  I will spend some time playing around with the IDE so that I have a better understanding of how programs are created and compiled.  I still need to read your 1200 page book I guess.  

Thank you again for responding.  I would have probably banged my head against a wall for a while until I realized that I had the wrong program and libraries.
